cpp/src/strings/convert/convert_durations.cu

```diff
@@ -0,0 +1,650 @@
+ * format_items.
+ *
+ * The vector of format_items are used when parsing a string into duration
+ * components and when formatting a string from duration components.
+ */
+struct format_compiler {
+  std::string format;
+  rmm::device_vector<format_item> d_items;
+
+  std::map<char, int8_t> specifier_lengths = {{'d', -1},
+                                              {'+', -1},  // only for negative days
+                                              {'H', 2},
+                                              {'M', 2},
+                                              {'S', 2},
+                                              {'u', -1},   // 0 or 6+1(dot)
+                                              {'f', -1}};  // 0 or <=9+1(dot) without trialing zeros
+
+  format_compiler(const char* format) : format(format) {}
+
+  format_item const* compile_to_device()
+  {
+    std::vector<format_item> items;
+    const char* str         = format.c_str();
+    auto length             = format.length();
+    const bool is_isoformat = str[0] == 'P';
+    if (is_isoformat) {
+      specifier_lengths['H'] = -1;
+      specifier_lengths['M'] = -1;
+      specifier_lengths['S'] = -1;
+    }
+    while (length > 0) {
+      char ch = *str++;
+      length--;
+      if (ch != '%') {
+        items.push_back(format_item::new_delimiter(ch));
+        continue;
+      }
+      CUDF_EXPECTS(length > 0, "Unfinished specifier in duration format");
+
+      ch = *str++;
+      length--;
+      if (ch == '%')  // escaped % char
+      {
+        items.push_back(format_item::new_delimiter(ch));
+        continue;
+      }
+      if (ch >= '0' && ch <= '9') {
+        CUDF_EXPECTS(*str == 'f', "precision not supported for specifier: " + std::string(1, *str));
+        specifier_lengths[*str] = static_cast<int8_t>(ch - '0') + 1;  // +1 is for dot
+        ch                      = *str++;
+        length--;
+      }
+      CUDF_EXPECTS(specifier_lengths.find(ch) != specifier_lengths.end(),
+                   "invalid format specifier: " + std::string(1, ch));
+
+      int8_t spec_length = specifier_lengths[ch];
+      items.push_back(format_item::new_specifier(ch, spec_length));
+    }
+    // create program in device memory
+    d_items.resize(items.size());
+    CUDA_TRY(cudaMemcpyAsync(
+      d_items.data().get(), items.data(), items.size() * sizeof(items[0]), cudaMemcpyHostToDevice));
+    return d_items.data().get();
+  }
+
+  // these calls are only valid after compile_to_device is called
+  size_type items_count() const { return static_cast<size_type>(d_items.size()); }
+};
+
+__device__ void dissect_duration(int64_t duration, int32_t* timeparts, type_id units)
+{
+  if (units == type_id::DURATION_DAYS) {
+    timeparts[DU_DAY] = static_cast<int32_t>(duration);
+    return;
+  }
+
+  // Specialized modulo expression that handles negative values.
+  // Examples:
+  //     modulo(1,60)    1
+  //     modulo(-1,60)  59
+  auto modulo_time = [](int64_t time, int64_t base) {
```

Comment:
This is needed to handle negative values for time units as shown in the comment.

Reply:
HH:MM:SS.9f are all in positive values. days are only negative values in string format.

Reply:
The comment doesn't explain what it actually _means_.

What is the meaning of `-1` days? Why is it equal to 59 days?

Reply:
`-1 minute` is `-1 days +23:59:00`. So, -1 will be converted to 59. 
In string representation, hours, minutes, seconds, subseconds take only positive values. (correspondingly, days get adjusted). (pandas timedelta format discussion https://github.com/pandas-dev/pandas/issues/17232)

Reply:
> -1 minute is -1 days +23:59:00

That's Pandas nonsense. Based on that issue, looks like I'm not the only one who thinks so. 

libcudf is _not_ Pandas and we should not beholden to matching exactly what they do. 

We should be following what's outlined here as closely as possible: https://en.cppreference.com/w/cpp/chrono/duration/formatter

Reply:
Big +1 to what Jake said above. In general when we print data in cudf python, we send a minimal amount of data to the host and allow Pandas to format the data for us to make sure we follow its semantics. Implementing the print formatting of Pandas on the GPU is completely out of scope.

Reply:
I agree too. I updated to stick to `std::format`. I considered pandas format for string conversion because these APIs will be useful for `write_csv` and `read_csv` API of durations column. Similar approach is taken to use `from_timestamps` in csv write code.
For `write_csv` of duration columns, should new APIs be implemented that can parse _pandas timedelta format_?